### DISCUSSION TITLE
Build both sdist and wheel (rather than one or the other)

### DISCUSSION THREAD

**Discussion by mcarans** [tox-dev/tox#2911] [Q&A] [answered]
## What's the problem this feature will solve?
I would like to have built both the sdist and wheel. 

## Describe the solution you'd like
Option to build both sdist and wheel

I have been using wheel because that's what I want to put on pypi but turns out I need sdist for conda.

This is my tox setup:

```
legacy_tox_ini = """
[tox]
envlist = py311, lint

[gh-actions]
python =
    3: py311, lint

[testenv]
recreate = true
package = wheel
wheel_build_env = .pkg
deps =
    -r test-requirements.txt
commands =
    pytest --cov=hdx --no-cov-on-fail --junitxml=.tox/test-results.xml --cov-report=xml --cov-report=term-missing

[testenv:lint]
deps =
    flake8
    flake8-isort
    flake8-black
    flake8-pyproject
commands =
    flake8 --color=always src tests

[testenv:docs]
deps =
    pydoc-markdown
commands =
    pydoc-markdown

[testenv:publish]
pass_env = SSH_AUTH_SOCK, TWINE_USERNAME, TWINE_PASSWORD
deps =
    twine

commands =
    twine upload {work_dir}/{package_env}/dist/*
```

**Comment by gaborbernat** [comment, marked as answer]
You can already do this. Create a different env for sdist; e.g.

```
[testenv:py311]
package = wheel

[testenv:py311-sdist]
package = sdist
```
Note setting the `recreate = true`  flag is not needed anymore with tox 4.

**Reply by mcarans** [reply]
Thanks. Adding `package = sdist` to `[testenv:publish]` worked. When I called, `tox -e publish`, the wheel was built by the `[testenv]` step while the sdist was built by the `[testenv:publish]` step (which then uploaded both to pypi). 

It seems that although `[testenv:publish]` inherits from `[testenv]`, `package = sdist` in `[testenv:publish]` does not override `package = wheel` in `[testenv]` but runs afterwards which works well for my use case.

**Reply by gaborbernat** [reply]
There's no testenv step. Likely you're running another env that builds the wheel.

**Reply by mcarans** [reply]
That is odd. I am running this in GitHub Actions. As you can see in this [Github Actions workflow](https://github.com/OCHA-DAP/hdx-python-utilities/blob/main/.github/workflows/publish.yml), I only call `tox -e publish`.

In the [log output](https://github.com/OCHA-DAP/hdx-python-utilities/actions/runs/4060324074/jobs/6989298480) of GitHub Actions, you can see that it has:
```
ROOT: tox-gh-actions won't override envlist because envlist is explicitly given via TOXENV or -e option
...
.pkg: get_requires_for_build_wheel> python /opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: install_requires_for_build_wheel> python -I -m pip install wheel
.pkg: get_requires_for_build_sdist> python /opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
...
.pkg: build_wheel> python /opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
```

Hence later it was able to upload both wheel and sdist to pypi:
```
Uploading hdx_python_utilities-3.5.4-py2.py3-none-any.whl
...
Uploading hdx-python-utilities-3.5.4.tar.gz
```
